kythe/docs/rfc/2967.md

```diff
@@ -0,0 +1,64 @@
+
+
+## Background {#background}
+
+Currently, the Kythe graph largely assumes that it is built for a single configuration, at a consistent revision for a particular corpus.  Historically, this has worked well.  But as time marches onward and the diversity of platforms built from a single corpus has grown it has left a widening gap in the information available to users with only partial cross references for some files (due to conditional compilation) or other files entirely omitted due to build-time selection.  In order to work around this lack, some third-party corpora (most notably Chromium) have begun to use the "root" VName field to encode this information.  Similarly, the extant "prodkernel" corpus has need of this, but currently makes do without. We reasonably anticipate that other soon-to-be-corpora such as Android and Tensorflow would encounter similar limitations.
+
+
+## Overview {#overview}
+
+Rather than introduce any new VName fields, nodes or edges build configuration can be adequately represented via a signature-relevant fact on anchor nodes and some ancillary clarifications to the guidelines for root and corpus.  This is predicated on:
+
+
+1.  Non-generated source files exist independent of build configuration.
+1.  Nodes must already incorporate relevant information into their signature.  This includes any elements related to build-configuration which materially alters the represented entity. This is currently already done by non-anchor (aka "semantic") nodes, but build configuration has not been considered relevant for anchors.
+1.  Certain languages will necessarily emit different anchors under different build configurations (C/C++ are particularly subject to this).
+1.  Generated source files must incorporate the build configuration into root in some fashion, if relevant.
+1.  The relevant information should be in a form which is readily available for filtering and display.
+
+The end result is that build configuration is most tightly bound to anchors both semantically and practically.  This design also allows for including references to entities regardless of build configuration as well as filtering only references for a particular set.  Indexers for which build configuration is irrelevant may omit the fact.
+
+
+## Detailed Design {#detailed-design}
+
+
+### Graph {#graph}
+
+The Kythe graph will be augmented such that:
+
+
+1.  File nodes with a non-empty root VName field are explicitly considered to have been generated by the build.  If the contents of generated files differ with the build configuration, the root must include a unique string per build configuration.  This follows from the general requirements that unique graph entities be given unique VNames.  This allows clients to trivially know if they should request the source for a particular file from the underlying repository or from Kythe.
```

Comment:
Here I think you've touched on an interesting idea: The distinction between "parts of the name related to output" (e.g., code generation) vs "parts of the name related to input" (e.g., configuration settings). Does that separation buy us something?

Reply:
Perhaps worth ruminating on, but won't impact this particular design, IMO.